### ISSUE TITLE
remove dependencies on {Sim} sim in numerous constructors

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/joist#188]
Too many components in joist have a {Sim} sim constructor parameter, allowing them to access everything in Sim.  Examples:

> function HomeScreen( sim, options )
> function NavigationBar( width, height, sim, screens, model )
> function NavigationBarScreenButton( sim, screen, navBarHeight, minWidth )
> function PhetButton( sim, options ) 
> function SettingsDialog( sim )
> function AboutDialog( sim, Brand )

This looks like quick-&-dirty prototype code, not production code. These are unnecessarily-wide interfaces, allowing coupling that shouldn't occur. And there are already cases where this is messed up.  For example:

Here's how Sim.js creates the NavigationBar:

> 335 sim.navigationBar = new NavigationBar( navbarWidth, navbarHeight, sim, screens, sim.simModel );

NavigationBar then creates NavigationBarScreenButtons:

> 62 return new NavigationBarScreenButton( sim, screen, thisNode.navBarHeight, 0 );

NavigationScreenButton doesn't use the model that was passed to NavigationBar, it uses sim.simModel:

> 36 var selected = sim.simModel.screenIndexProperty.valueEquals

In this case, we're just lucky that Sim happens to be passing sim.simModel to NavigationBar.

**Comment by samreid** [comment]
I'm interested in fixing this once we switch to joist/ohtwo so I don't have to do this twice.

**Comment by samreid** [comment]
Is this what you had in mind? 1e6d282 Or do you want to bundle the information into an object literal?  Should the credits be in a new options parameter since it is apparently optional?

**Comment by samreid** [comment]
HomeScreen & HomeScreenView use so many parts of sim that passing in the entire sim instance may be warranted.

**Comment by samreid** [comment]
Likewise, PhetMenu.js is making 11 distinct API calls on Sim.js (some more than once).  Does this mean we should pass in 11 separate arguments into the PhetMenu constructor?  I suspect not, but just want to ping @pixelzoom and find out where he draws the line, or whether there is another solution.

**Comment by samreid** [comment]
I removed sim from many places.  It is still in the following places, which may be warranted.  Let's discuss it:

1. HomeScreen -> HomeScreenView
2. PhetButton
3. NavigationBar
4. PhetMenu

Leaving assigned to @pixelzoom for discussion.

**Comment by pixelzoom** [comment]
I realize that it's been awhile, and I haven't gotten to this. If I don't get to it before @samreid's leave, I'll handle it.

**Comment by pixelzoom** [comment]
No plans to work on this in the near future, so unassigning for now.

**Comment by pixelzoom** [comment]
No plans to work on this in the near future, so unassigning for now.

**Comment by pixelzoom** [comment]
Reviewed... This is still a problem, and may have gotten worse.  In addition to passing the complete `{Sim} sim` instance to components, some components are now using `window.phet.joist.sim`.

**Comment by samreid** [comment]
One new occurrence of this pattern is A11yButtonsHBox.js, which takes `Sim` as a constructor parameter, and accesses:
* sim.soundPartOfTheAPI
* sim.supportsSound
* sim.accessibilityPartOfTheAPI
* sim.hasKeyboardHelpContent
* sim.screenProperty
* sim.supportsInteractiveDescription

What is a better pattern here?

Also, there are now around 300 occurrences of `phet.joist.sim` for things like:
* phet.joist.sim.isSettingPhetioStateProperty
* phet.joist.sim.frameStartedEmitter
* phet.joist.sim.utteranceQueue
* phet.joist.sim.selfVoicingUtteranceQueue

What are some alternatives?

**Comment by pixelzoom** [comment]
Sim is one monolithic "flat" class that has too many responsibilities.  It's a dumping ground for features. Need to add something to the sim quickly? Add a field to Sim.  Sim hasn't been designed, it's been grown.

So my advice is to design Sim.  A good first step would be to divide Sim's responsibilities up amongst smaller "delegate" classes.  Then (for example) something that needs to deal with sound doesn't need to be given access to the entire Sim API, it can just be passed the subcomponent that is responsible for sound.  Sound and A11y features look like excellent candidates for delegate classes.  I count at least 6 fields that have been added to Sim for A11y. 

You've also pointed out that sometimes Sim is passed as a constructor/method parameter, and sometimes it's accessed as a global.  It's very concerning that there are ~300 (!) occurrences of `phet.joist.sim`.  Does PhET really have that many features that can't function without globals? Were globals chosen because there was no other way, or was it a short-cut to get something done quickly? I suspect the latter in most cases.

**Comment by zepumph** [comment]
Agreed! I want to also tag https://github.com/phetsims/joist/issues/208, an old issue that is (ish) a duplicate.

**Comment by zepumph** [comment]
>Does PhET really have that many features that can't function without globals?

Part of that is being worked on over in https://github.com/phetsims/joist/issues/686. The brunt of the issue is that most common code shouldn't depend on joist, so instead of importing the modules, from it, we use a global to get the sim instance. That issue is to help break that dependency when possible, and provide flexibility when it is not possible to break entirely. 

I'm not sure what to do with this issue right now; what makes it high priority? It seems like the time to tackle a division of Sim reponsibilities is in great Joist 3.0 work in https://github.com/phetsims/joist/issues/678 

@samreid does that seem right to you?

**Comment by samreid** [comment]
Some comments:

1. We could count the number of occurrences of `phet.joist.sim.isSettingPhetioStateProperty` vs `phet.joist.sim.utteranceQueue` and see what else is being used.
2. Even after https://github.com/phetsims/joist/issues/686 is complete, it will still entail (guarded) access to joist globals.
3. Would it help to split `utteranceQueue` and `isSettingPhetioStateProperty` out of `sim` so that common code could access it directly without going through `sim`?
4. I'll reduce to normal priority.

**Comment by zepumph** [comment]
>Would it help to split utteranceQueue and isSettingPhetioStateProperty out of sim so that common code could access it directly without going through sim?

There was difficulty when originally trying to do that with utteranceQueue, though it would have been preferred. Please see https://github.com/phetsims/utterance-queue/issues/13 for work in that regard.

**Comment by samreid** [comment]
There is a proposal in https://github.com/phetsims/utterance-queue/issues/13#issuecomment-780989547 for how that may reduce usage of `phet.joist.sim`.  I'm not planning further work on this issue at the moment, unassigning for now.

**Comment by zepumph** [comment]
While working on another issue, I saw that PhetioEngine.onSimConstructionStarted takes the whole sim, but it doesn't need it. I'll refactor that now.
